netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -258,51 +297,132 @@ public void operationComplete(ChannelFuture future) throws Exception {
         ch.closeFuture().addListener(loopReleaser);
       }
     });
-    // Bind and start to accept incoming connections.
-    ChannelFuture future = b.bind(address);
-    // We'd love to observe interruption, but if interrupted we will need to close the channel,
-    // which itself would need an await() to guarantee the port is not used when the method returns.
-    // See #6850
-    future.awaitUninterruptibly();
-    if (!future.isSuccess()) {
-      throw new IOException(String.format("Failed to bind to address %s", address), future.cause());
+
+    Future<Map<ChannelFuture, SocketAddress>> bindCallFuture =
+        bossExecutor.submit(new Callable<Map<ChannelFuture, SocketAddress>>() {
+          @Override
+          public Map<ChannelFuture, SocketAddress> call() {
+            Map<ChannelFuture, SocketAddress> map = new HashMap<>();
+            for (SocketAddress address: addresses) {
+                ChannelFuture future = b.bind(address);
+                map.put(future, address);
+            }
+            return map;
+          }
+        }
+    );
+    Map<ChannelFuture, SocketAddress> channelFutures =
+        bindCallFuture.awaitUninterruptibly().getNow();
```

Comment:
This produces the wrong results if the future hasn't failed and just hasn't completed yet.

Reply:
Could you help elaborate on why awaitUninterruptibly() is not enough to have the result ready when getNow() is called? Thanks.

Reply:
I think I was probably blind to the `awaitUninterruptibly()`. This looks fine.